include/wil/cppwinrt_authoring.h

```diff
@@ -0,0 +1,247 @@
+        {
+            m_propertyChanged.remove(token);
+        }
+        
+        Type& self()
+        {
+            return *static_cast<Type*>(this);
+        }
+
+        /**
+         * @brief Raises a property change notification event
+         * @param name The name of the property
+         * @return
+         * @details Usage example\n
+         * C++
+         * @code
+         * void MyPage::DoSomething()
+         * {
+         *     // modify MyInt
+         *     // MyInt = ...
+         *
+         *     // now send a notification to update the bound UI elements
+         *     RaisePropertyChanged(L"MyInt");
+         * }
+         * @endcode
+        */
+        auto RaisePropertyChanged(std::wstring_view name) {
+            return m_propertyChanged(self(), Xaml_Data_PropertyChangedEventHandler{ name });
+        }
+    protected:
+        winrt::event<Xaml_Data_PropertyChangedEventHandler> m_propertyChanged;
+    };
+
+    /**
+     * @brief Implements a property type with notifications
+     * @tparam T the property type
+     * @details Use the #INIT_NOTIFY_PROPERTY macro to initialize this property in your class constructor. This will set up the right property name, and bind it to the `notify_property_changed_base` implementation.
+    */
+    template<typename T,
+        typename Xaml_Data_PropertyChangedEventHandler = wil::details::Xaml_Data_PropertyChangedEventHandler,
+        typename Xaml_Data_PropertyChangedEventArgs = wil::details::Xaml_Data_PropertyChangedEventArgs>
+    struct single_threaded_notifying_property : single_threaded_rw_property<T>
+    {
+        using Type = T;
+
+        using single_threaded_rw_property<T>::operator();
+
+        void operator()(const T& value)
+        {
+            if (value != this->m_value)
+            {
+                single_threaded_rw_property<T>::operator()(value);
+                if (m_npc)
+                {
+                    (*m_npc)(m_sender, Xaml_Data_PropertyChangedEventArgs{ m_name });
+                }
+            }
+        }
+
+        template<typename... TArgs>
+        single_threaded_notifying_property(
+            winrt::event<Xaml_Data_PropertyChangedEventHandler>* npc,
+            winrt::Windows::Foundation::IInspectable sender,
+            std::wstring_view name,
+            TArgs&&... args) :
+            single_threaded_rw_property<T>(std::forward<TArgs...>(args)...),
+            m_name(name),
+            m_npc(npc),
+            m_sender(sender)
+        {}
+
+        single_threaded_notifying_property(const single_threaded_notifying_property&) = default;
+        single_threaded_notifying_property(single_threaded_notifying_property&&) = default;
+        std::wstring_view Name() const noexcept { return m_name; }
+    private:
+        std::wstring_view m_name;
+        winrt::event<Xaml_Data_PropertyChangedEventHandler>* m_npc;
+        winrt::Windows::Foundation::IInspectable m_sender;

```

Comment:
Isn't the sender typically `this`? So wouldn't this keep a circular strong reference?

Reply:
`this` in this context would refer to the property. The sender is the object which hosts the property (e.g. a Panel, that has a property `Width`, when the property changes, the sender is the panel, not the `Width` property)

Reply:
Sorry, when I say "this" I'm referring to the object hosting the property. E.g. if class `Foo` has a member of type `single_threaded_notifying_property`, then my understanding is that the sender (and what I call "this") is the `Foo` object.

If so, that's a circular reference. The `Foo` object has a member `single_threaded_notifying_property`, which is holding a strong reference to the `Foo` object

Reply:
the latest iteration already converted it to be a weak_ref